eclipse-collections-api/src/main/java/org/eclipse/collections/api/factory/list/Lists.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2017 Goldman Sachs and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v. 1.0 which accompany this distribution.
+ * The Eclipse Public License is available at http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ */
+
+package org.eclipse.collections.api.factory.list;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ServiceLoader;
+import java.util.stream.Collectors;
+
+/**
+ * This class should be used to create instances of MutableList, ImmutableList and FixedSizeList
+ * <p>
+ * Mutable Examples:
+ *
+ * <pre>
+ * MutableList&lt;String&gt; emptyList = Lists.mutable.empty();
+ * MutableList&lt;String&gt; listWith = Lists.mutable.with("a", "b", "c");
+ * MutableList&lt;String&gt; listOf = Lists.mutable.of("a", "b", "c");
+ * </pre>
+ *
+ * Immutable Examples:
+ *
+ * <pre>
+ * ImmutableList&lt;String&gt; emptyList = Lists.immutable.empty();
+ * ImmutableList&lt;String&gt; listWith = Lists.immutable.with("a", "b", "c");
+ * ImmutableList&lt;String&gt; listOf = Lists.immutable.of("a", "b", "c");
+ * </pre>
+ *
+ * FixedSize Examples:
+ *
+ * <pre>
+ * FixedSizeList&lt;String&gt; emptyList = Lists.fixedSize.empty();
+ * FixedSizeList&lt;String&gt; listWith = Lists.fixedSize.with("a", "b", "c");
+ * FixedSizeList&lt;String&gt; listOf = Lists.fixedSize.of("a", "b", "c");
+ * </pre>
+ */
+@SuppressWarnings("ConstantNamingConvention")
+public final class Lists
```

Comment:
This is a *copy* of Lists with changes. The main copy lives in impl, this one lives in api. I'll mark the main differences below.

Reply:
Are you proposing to have two factories or remove the impl one eventually?

Reply:
Good question. I could see deprecating the impl one and eventually deleting it. It's pretty disruptive to import statements though and pretty widely used. I could see keeping both, but then the new ones should probably have different names (ListFactory instead of Lists, or something like that). The downside is that the old one has the better name, and the good one is the one we'd encourage to use. What do you think?

Reply:
I'm not convinced having two is a bad thing. The impl has the advantage that it never throws. Maybe if we play our cards right, we can get the api one to call the impl one, so that the api one exists for the default method implementation and the use cases where another framework doesn't want a dependency on the impl at compile time (with runtime being potentially optional).